HP-1796: UI facelift

NikoHelle opened the pull request:
Changed styles according to the [new design](https://share.goabstract.com/ef4b8648-6b86-41b9-b1c2-f0c78721a436?collectionLayerId&mode).

Also cleaned up the codebase. Removed "multi-item"-related components, because those have been obsolete for a while. A lot of code was merged to make components simpler and after that it was easier to make style changes, because container elements were unified and common styles easier to use.

This branch is currently in https://profiili.dev.hel.ninja/

Commit:
(HP-1796) Fix all heading styles to match layout and HDS

Also changed the max-width of the login-box, because new font-sizes do not fit well. Max-width is the same as in AccessibilityStatement

Commit:
(HP-1796) New heading style "hero"

Profile and ServiceConnections pages should have this new style. Not other pages.

The style is created with ":before" because the heading is deep inside parent containers and the page wide background colour would not be possible otherwise.

Except deconstructing all parent containers and/or their padding.

Commit:
(HP-1796) New styled boxes with golden border

Commit:
(HP-1796) Added second close button to Expanding panel

Now it matches the HDS Accordion and new designs

Commit:
(HP-1796) Change translations

Commit:
(HP-1796) Re-style service connections boxes

close-button-container position changed to absolute so it won't take space and widen the gap below delete button

Commit:
(HP-1796) Group phonenumber and email editor under "Contact"

First MultiItemEditor was stripped of container elements.

Then added AddressEditor to kind of revert that, but make component look the same as before.

Mostly restyled headings and some borders.

Commit:
(HP-1796) Change button to look like hel.fi buttons

Easiest way was to override css vars via a wrapper component.

Commit:
(HP-1796) Fix error notification padding in ServiceConnections

The notification was not wrapped like other elements

Commit:
(HP-1796) Move editing tools to a hook from multiItemEditor

There can be only one address / phone number. That has been changed before.

The components for both dataTypes were very similar before, but now it is better to remove the multiItem editor which handles both.

Components are now more unique and harder to keep them logically in one, so first move functionalities to a hook and then split the component.

The hook is easier to re-use later.

Commit:
(HP-1796) Moved common functionalities to useCommonEditHandler hook

Creating one hook for editing.

Also removed looping through data items. Only one or zero can exist.

Commit:
(HP-1796) Simplified actions

Edit handling is spread to multiple places and confusing. Simplified it by moving all handling to single hook.

Created an AddButton component that uses same hook so props do not have to be parsed and passed to a Button.

Commit:
(HP-1796) Use the useCommonEditHandling also with email editing

The save action must return {success: true/false} with emails, so the verification notification can be shown.

Commit:
(HP-1796) Use the useCommonEditHandling also with basic data

Added also generic save prevention when saving same data as before. Save success message is shown, so user is not confused.

Commit:
(HP-1796) Move content from MultiItemEditor to actual editor components

No need to keep moved data in MultiItemEditor.

Also easier to handle elements and their styling when everything is in the editor component and not partly in the parent.

Commit:
(HP-1796) Unify container styles

Css classNames were confusing and usage was illogical.

Same kind of elements had different classNames which makes restyling difficult.

Styles had a lot of repetition, different classes doing same things.

Merged styles and forced elements to use same classes.

All components use same containers in same places, so it is easier to change styles everywhere.

EditButtons buttonClassNames props was not used anymore so removed it.

Removed also unused css modules

Commit:
(HP-1796) Same container styles to Download/Delete components

Commit:
(HP-1796) Adjust element spacings

DeleteProfile.module.css became obsolete.

Commit:
(HP-1796) VerifiedPersonalInformation should have bottom border

I should also have no horizontal padding

Commit:
(HP-1796) Horizontal lines in content boxes only with texts

Commit:
(HP-1796) Adjust mobile spacings

Commit:
(HP-1796) Fix link to Service Connections

Switched to use Link from react-router-dom so the page won't reload page when clicked.

Commit:
(HP-1796) Restyle Expanding panel buttons

Added a new button variant 'transparent' to StyledButton

For some reason styles in the production build have different order and styling did not take effect in StyledButtons at all

Increased the priority of custom styles by adding "button" to selector

Commit:
(HP-1796) Change title of the Service connections page.

The old one was overly long

Commit:
(HP-1796) Updated caniuse-lite

While building Browserslist keeps complaining "caniuse-lite is outdated"

Updated it.

Yarn also re-ordered some packages in package.json

Commit:
(HP-1796) Fix skewed sides of a bottom border

Having side borders with thick bottom border makes the sides of the bottom border skewed.

Commit:
(HP-1796) Grouped editable data and added a new heading

When user has verified personal information, there is a new heading  before the editable fields. Data sets are grouped into regions.

Adding regions is better than swapping heading levels. Verified data is in one region and editable in another.

The heading is not h2, or headings beneath it would have to change to h3. But conditionally - only when user has verified data. Or hierarchial order of heading elements would break.

The heading, a span element, is hidden from screen readers, because region label has same text.

Splitted editable sections to EditableProfileData, so ProfileInformation is cleaner and new logic is in the new component.

Removed now obsolete "bottomBorder" -prop from ProfileSection.

h2 headings are visually too large, so made them look like h3.

Commit:
(HP-1796) Remove borders from download and delete profile

These components should look like they were before

Commit:
(HP-1796) Same styling to CreateProfile

Remove many unused styles too.

Commit:
(HP-1796) Added phoneNumberEditor component

Makes EditableProfileData cleaner and MultiItemEditor can be removed later.

Commit:
(HP-1796) Moved contents of the MultiItemEditor to actual components

MultiItemEditor is not needed anymore as a wrapper component. It is very simple and therefore not needed as a wrapper.

Commit:
(HP-1796) Moved AddButton to own folder

Commit:
(HP-1796) Removed MultiItemEditor

Moved individual editor components under Address/PhoneNumberEditor and renamed them.

Changed tests to use main components Address/PhoneNumberEditor

MultiItemEditor is not needed anymore

Commit:
(HP-1796) Mobile view adjustments

Expanding buttons are too wide in mobile. HDS does not shown labels in Accordions, so those can be hidden in mobile.

Text fields were also too tightly aligned in mobile.

Commit:
(HP-1796) Fix heading shifting in Phone and Email editors

When switching to edit mode, the heading shifts few pixels.

Also re-arranged h3 in phone editing to match order in email editing.

No need to shown noPhone translation when adding new one, so removed it from there

akikoskinen reviewed (commented):
Overall looks nice. I see some differences between the implementation and the design. Are these intentional, or should I point them out?

Also this one error view could perhaps use some style tuning:

<img width="871" alt="Screenshot_20230306_113507" src="https://user-images.githubusercontent.com/758558/223073764-160fd029-1103-4872-98a3-79984c98e853.png">

NikoHelle commented:
I'll have to make some changes to the buttons. Otherwise implementation should be ok, at least designer did not say anything. The design is based on hel.fi, but we use HDS, so they cannot match 100%.

I'll add padding to the error too.

Commit:
(HP-1796) Adjust delete/download profile spacings

Created a common class for spacing child elements.

Fixes also child padding problems via .uneditable-box-content

Commit:
(HP-1796) Fix SonarCloud duplicate code errors

akikoskinen reviewed (changes requested):
Not sure if this PR is the reason for this behavior, but I noticed the following. Locally I shut down Profile API, so that Profile UI can't save any profile data changes. When I edit some data and try to save, an error message is shown correctly. When I then click the save button again, Profile UI doesn't even try to send the data to Profile API, it just informs that saving was successful.

akikoskinen reviewed (commented):
Here's a collection of some error messages. While they all look nice on their own, they all look different. Would it be important to use a unified style for the error messages, or is that not a goal?

<img width="855" alt="Screenshot_20230308_083047" src="https://user-images.githubusercontent.com/758558/223637395-35fe350d-4225-48e2-b92e-f9a311d79b1b.png">
<img width="899" alt="Screenshot_20230308_083108" src="https://user-images.githubusercontent.com/758558/223637404-f0eb91d3-dc95-4473-b381-20c62fac0ff6.png">
<img width="336" alt="Screenshot_20230308_083331" src="https://user-images.githubusercontent.com/758558/223637411-fa758452-06ff-4c52-a41e-fcde23ffa091.png">
<img width="452" alt="Screenshot_20230308_083406" src="https://user-images.githubusercontent.com/758558/223637414-e4de9eec-68f6-4b45-8630-cd499eeb98a8.png">
<img width="564" alt="Screenshot_20230308_083452" src="https://user-images.githubusercontent.com/758558/223637418-0bcbca82-1ba0-4f0b-bd90-2a59f02199ae.png">

Commit:
(HP-1796) Added new tests for AddressEditor

The editor should not save and post unedited values. Same functionality was already in emailEditor.

Also a new test for a bug when saving same data for a second time after an error.

Second attempt was not done, because data was not changed. Save attempt resulted in save success message.

NikoHelle commented:
Great summary of errors! Thx!

Here are fixed ones

![image](https://user-images.githubusercontent.com/2639230/223796425-82be37ef-f1a9-4ed4-995c-40af0ba672e3.png)

And success messages
![image](https://user-images.githubusercontent.com/2639230/223797258-a42379e8-9b6f-43a6-80f8-5ce27aaca760.png)

Commit:
(HP-1796) Harmonize editing result notifications

There was few different styles. Now one for success and one for error notifications

Errors are always shown above the buttons, so they won't possibly be placed below the fold - especially on mobile.

Success messages are shown on the bottom of the component.

Notifications are shown in many places and some of those require different spacings so added variants as props to the component.

NikoHelle commented:
> Not sure if this PR is the reason for this behavior, but I noticed the following. Locally I shut down Profile API, so that Profile UI can't save any profile data changes. When I edit some data and try to save, an error message is shown correctly. When I then click the save button again, Profile UI doesn't even try to send the data to Profile API, it just informs that saving was successful.

This is also in the old version, produceable in the test env. But only with emails, where data is not sent, if new email matched the old one. Same feature was added to other fields in this pr. The bug is fixed now and tests were added to prevent it.

akikoskinen commented on src/profile/components/phoneNumberEditor/PhoneNumberFormAndData.tsx:
```diff
@@ -0,0 +1,264 @@
+    editButtonId,
+    removeButtonId,
+    dataType,
+    noticationContent,
```
Typo: `noticationContent`.

NikoHelle replied:
Fixed - also in all files where used

akikoskinen reviewed (changes requested):
Only one commit squashing and a typo left to fix.

The commit squashing was [commented into the extra commit](https://github.com/City-of-Helsinki/open-city-profile-ui/commit/794543061894899cb4690b85bfa39261397502ee).

Commit:
(HP-1796) Gather all UI tests to one file

All UI test follow the same pattern now when there are no multiple items.

User flow is almost same with all datatypes, so created common test flows.

Commit:
(HP-1796) Convert BasicData tests to new test flow

Commit:
(HP-1796) Convert Address tests to new test flow

Commit:
(HP-1796) Convert Email tests to new test flow

Commit:
(HP-1796) Convert Phonenumber tests to new test flow

NikoHelle commented:
SonarCloud complained again about duplicate code in test files. Ended up re-writing all editing tests.

Gathered all test scenarios to one file and use that is all editing tests. 

Editing tests follow same flow in all tests ,so it was quite convenient to gather them. 

Data and input validation and test initialisations differ, so that took same time to convert to common tests. 

Additions start at in [this commit](https://github.com/City-of-Helsinki/open-city-profile-ui/pull/287/commits/e2bb6b844ba2f9460fe2e65386cc3ad56701641c)

akikoskinen reviewed (approved):
:+1:
